ext/web/16_image_data.js

```diff
@@ -0,0 +1,174 @@
+
+    // Overload: new ImageData(data, sw [, sh [, settings ] ])
+    if (typeof arg0 === "object") {
+      data = arg0;
+      sourceWidth = typeof arg1 !== "undefined"
+        ? parseInt(arg1, 10)
+        : undefined;
+      sourceHeight = typeof arg2 !== "undefined"
+        ? parseInt(arg2, 10)
+        : undefined;
+      settings = arg3;
+
+      if (!(data instanceof Uint8ClampedArray)) {
+        throw new TypeError(
+          "Failed to construct 'ImageData': The provided value is not an instance of Uint8ClampedArray.",
+          "TypeError",
+        );
+      }
+
+      if (Number.isNaN(sourceWidth) || sourceWidth < 1) {
+        throw new DOMException(
+          "Failed to construct 'ImageData': The source width is zero or not a number.",
+          "IndexSizeError",
+        );
+      }
+
+      if (
+        typeof sourceHeight !== "undefined" && Number.isNaN(sourceWidth) ||
+        sourceHeight < 1
+      ) {
+        throw new DOMException(
+          "Failed to construct 'ImageData': The source height is zero or not a number.",
+          "IndexSizeError",
+        );
+      }
+
+      if (data.length % 4 !== 0) {
+        throw new DOMException(
+          "Failed to construct 'ImageData': The input data length is not a multiple of 4.",
+          "InvalidStateError",
+        );
+      }
+
+      if (data.length / 4 % sourceWidth !== 0) {
+        throw new DOMException(
+          "Failed to construct 'ImageData': The input data length is not a multiple of (4 * width).",
+          "IndexSizeError",
+        );
+      }
+
+      if (
+        typeof sourceHeight !== "undefined" &&
+        (sourceWidth * sourceHeight * 4 !== data.length)
+      ) {
+        throw new DOMException(
+          "Failed to construct 'ImageData': The input data length is not equal to (4 * width * height).",
+          "IndexSizeError",
+        );
+      }
+
+      if (
+        typeof settings !== "undefined" &&
+        typeof settings.colorSpace !== "undefined" &&
+        !PredefinedColorSpace.includes(settings.colorSpace)
+      ) {
+        throw new TypeError(
+          `Failed to read the 'colorSpace' property from 'ImageDataSettings': The provided value '${settings.colorSpace}' is not a valid enum value of type PredefinedColorSpace.`,
+        );
+      }
+
+      this.#width = sourceWidth;
+      this.#height = typeof sourceHeight === "undefined"
+        ? data.length / 4 / sourceWidth
+        : sourceHeight;
+      this.#data = data;
+      this.#colorSpace = typeof settings !== "undefined" &&
+          typeof settings.colorSpace !== "undefined"
+        ? settings.colorSpace
+        : "srgb";
+      return;
+    }
+
+    // Overload: new ImageData(sw, sh [, settings])
+    sourceWidth = typeof arg0 !== "undefined" ? parseInt(arg0, 10) : undefined;
+    sourceHeight = typeof arg1 !== "undefined" ? parseInt(arg1, 10) : undefined;
+    settings = arg2;
+
+    if (Number.isNaN(sourceWidth) || sourceWidth < 1) {
```

Comment:
There is some repeated logic between both overloads, but I think the class is simpler to understand with the duplication.

Please let me know there are better or more preferred ways to refactor.